### ISSUE TITLE
Improve error message

### ISSUE THREAD

**Issue by bluzky** [edgurgel/solid#93]
Currently if parsing failed, solid the return message

```
Reason: expected end of string, line: xxx
```
It does not provide much detail so you don't know where to fix or what to fix.
It will be better if the error message can provide more details:
- Which file ( in case using render tag )
- Which line ( currently it return only the line of opening tags)
  For example 
```liquid
{% if 1 == 1 %}
Hi
how are you {{name
{% endif %}
```
and the line which error return is line 1, not line 3

**Comment by edgurgel** [comment]
Yeah definitely something we need to improve! I'm almost done with a major refactor of tags and custom tags so I would avoid any big change on the parser for now. Hoping to get this out tomorrow!

**Comment by edgurgel** [comment]
For the error messages we can probably play with https://hexdocs.pm/nimble_parsec/NimbleParsec.html#label/3 ? I've never used it but I hope it can help us?

For the file and line case:

Maybe the `render` tag implementation has to catch the error and change the reason here:

https://github.com/edgurgel/solid/blob/157d79966fd2769ff0a19127e9a6d9b295da655e/lib/solid/tag/render.ex#L44

So maybe we could change it to use `Solid.parse` and we tweak the reason to include which `filename` caused that and then raise the exception?

**Comment by bluzky** [comment]
I have played with line and offset from the error return by `NimbleParsec` but it does not return the correct line number of the actual error.
The line number is the open tag where I use `parsec`, not the line number of the nested tag.
I have raised a question on Elixir Forum, I hope that someone may suggest a direction.

**Comment by bluzky** [comment]
I have a solution but it's quite cumbersome. I keep parsing the remaining template until error line is 1. then I got  the exact blog which cause bug. Something like this

```elixir
def parse(text, opts \\ []) do
    parser = Keyword.get(opts, :parser, Solid.Parser)
    template_file = Keyword.get(opts, :template)

    case parser.parse(text) do
      {:ok, result, _, _, _, _} ->
        {:ok, %Template{parsed_template: result}}

      {:error, reason, remaining, _, {line, _}, _} ->
        {line, reason, remaining} =
          Enum.reduce_while(1..100, {line, reason, remaining}, fn _, {line, _, remaining} ->
            case parser.parse(remaining) do
              {:error, reason, remaining, _, {l, _}, _} ->
                if l == 1 do
                  {:halt, {line + l, reason, remaining}}
                else
                  {:cont, {line + l, reason, remaining}}
                end

              _ ->
                {:halt, {0, nil, ""}}
            end
          end)

        [template | _] = String.split(remaining, "\n", parts: 2)
        {:error, TemplateError.exception([reason, template_file, line, template])}
    end
  end
```

**Comment by edgurgel** [comment]
Interesting! It might be a good start then? Until we find a better solution I think this is better than what we have today. What do you think?

**Comment by bluzky** [comment]
after do more testing with this idea, for some cases it points to correct line which cause error, some other cases it's still wrong. So I think that we should try other solutions
